test/README.md

```diff
@@ -0,0 +1,179 @@
+script will cause them to be run in that use case.
+
+For local development / maintenance of the tests however, it is convenient to be able to run them
+in various manners which are described here.
+
+### Pre-Submit Script
+
+The [`presubmit-tests.sh`](./presubmit-tests.sh) script is the entry point for the tests run
+from Prow when a PR is created.
+
+By default, the script will run `build tests`, `unit tests` and `integration tests`.  If you
+only want to run one type of tests, you can run this script with corresponding flags such as...
+
+```
+./test/presubmit-tests.sh
+./test/presubmit-tests.sh --build-tests
+./test/presubmit-tests.sh --unit-tests
+./test/presubmit-tests.sh --integration-tests
+```
+
+The script will call the [`e2e-tests.sh`](./e2e-tests.sh) script if it exists.
+
+### E2E-Tests Script
+
+The [`e2e-tests.sh`](./e2e-tests.sh) script is the entry point for running all the e2e tests.
+
+By default, it will create a new GKE cluster in project `$PROJECT_ID`, start Knative Eventing,
+upload test images to `$KO_DOCKER_REPO`, and run the `conformance` and `e2e` tests. After the
+tests finishes, it will delete the cluster.  To skip cluster teardown for diagnostic purposes,
+"--skip-teardowns" can be added to the script arguments or to the initialize command in
+e2e-tests.sh directly.
+
+```
+# Complete Tests (Create & Initialize K8S Cluster)
+./test/e2e-tests.sh
+
+# Skip Cluster Creation & Install Knative Eventing Into Existing Empty K8S Cluster (kubeconfig)
+./test/e2e-tests.sh --run-tests
+
+# Skip Cluster Creation & Knative Eventing Setup - Use Existing K8S Cluster With Eventing Pre-Installed
+kubectl apply --filename https://github.com/knative/eventing/releases/download/v0.17.3/eventing.yaml
+./test/e2e-tests.sh --run-tests --skip-knative-setup
+
+# Prevent TearDown Of Tests In Order To Debug Failures
+./test/e2e-tests.sh --run-tests --skip-teardowns
+```
+
+Note that local clusters often do not have the resources to run 12 parallel tests (the default) as the
+tests each tend to create their own namespaces and dispatchers which require a number of cluster resources.
+For example, a local Docker cluster with 4 CPUs and 8 GB of RAM will probably be able to handle 6 at maximum.
+Be sure to adequately set the MAX_PARALLEL_TESTS variable before running this script on a local cluster.
+
+>Remember to provide the `kn-eventing-test-pull-secret` K8S Secret as described
+>in [Private Docker Repositories](#private-docker-repositories) when using an existing cluster.
+
+The script will base the version of Knative Eventing installed in the K8S Cluster on the
+current git branch.  If the current git branch is based off of a release branch, it will
+use the latest version of the corresponding Knative Eventing release branch.  If it is
+based off of /master then the latest /master Knative Eventing will be cloned/built/installed.
+Optionally you can set the following environment variable to force a specific Knative Eventing
+release...
```

Comment:
<!--mdfmt-bot-->
Format markdown:
```suggestion
The script will base the version of Knative Eventing installed in the K8S
Cluster on the current git branch. If the current git branch is based off of a
release branch, it will use the latest version of the corresponding Knative
Eventing release branch. If it is based off of /master then the latest /master
Knative Eventing will be cloned/built/installed. Optionally you can set the
following environment variable to force a specific Knative Eventing release...
```